src/ball_tree.jl

```diff
@@ -3,9 +3,9 @@
 # which radius are determined from the given metric.
 # The tree uses the triangle inequality to prune the search space
 # when finding the neighbors to a point,
-immutable BallTree{T <: AbstractFloat, M <: Metric} <: NNTree{T, M}
-    data::Matrix{T}                       # dim x n_points array with floats
-    hyper_spheres::Vector{HyperSphere{T}} # Each hyper sphere bounds its children
+immutable BallTree{V <: AbstractVector, N, T, M <: Metric} <: NNTree{V, M}
+    data::Vector{V}
```

Comment:
I'm sure `Vector{Vector{T}}` is slower than `Matrix{T}` (as would be `Vector{MVector{T}}`), so people will need to be comfortable to use an immutable static array for the point type.